crates/sqlite_store/src/store.rs

```diff
@@ -0,0 +1,1095 @@
+use rusqlite::{named_params, Connection, Error};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+use std::default::Default;
+use std::marker::PhantomData;
+use std::str::FromStr;
+use std::{fmt::Debug, path::Path};
+
+use bdk_chain::bitcoin::consensus::{deserialize, serialize};
+use bdk_chain::bitcoin::hashes::Hash;
+use bdk_chain::bitcoin::{BlockHash, Network, OutPoint, ScriptBuf, Transaction, TxOut, Txid};
+use bdk_chain::miniscript::{Descriptor, DescriptorPublicKey};
+use bdk_chain::{
+    indexed_tx_graph, keychain, local_chain, tx_graph, Anchor, Append, BlockId,
+    ConfirmationHeightAnchor, ConfirmationTimeHeightAnchor, DescriptorExt, DescriptorId,
+    PersistBackend,
+};
+
+use crate::{ChangeSet, ReadError, WriteError};
+
+const MIGRATIONS: &[&str] = &[
+    // schema version control
+    "CREATE TABLE version (version INTEGER) STRICT;",
+    "INSERT INTO version VALUES (1);",
+    // network is the valid network for all other table data
+    "CREATE TABLE network
+    (name TEXT UNIQUE NOT NULL)
+    STRICT;",
+    // keychain is the json serialized keychain structure as JSONB,
+    // descriptor_id is a sha256::Hash id of the descriptor string w/o the checksum,
+    // descriptor is the complete descriptor string,
+    // last active index is a u32
+    "CREATE TABLE keychain
+    (keychain BLOB PRIMARY KEY NOT NULL,
+     descriptor TEXT UNIQUE NOT NULL,
+     descriptor_id BLOB UNIQUE NOT NULL,
+     last_revealed INTEGER)
+    STRICT;",
+    // hash is block hash hex string,
+    // block height is a u32,
+    // previous block hash hex string
+    "CREATE TABLE block
+    (hash TEXT PRIMARY KEY NOT NULL,
+     height INTEGER NOT NULL)
+    STRICT;",
+    // txid is transaction hash hex string (reversed)
+    // whole_tx is a consensus encoded transaction,
+    // last seen is a u64 unix epoch seconds
+    "CREATE TABLE tx
+    (txid TEXT PRIMARY KEY NOT NULL,
+     whole_tx BLOB NOT NULL,
+     last_seen INTEGER)
+    STRICT;",
+    // Outpoint txid hash hex string (reversed)
+    // Outpoint vout
+    // TxOut value as SATs
+    // TxOut script consensus encoded
+    "CREATE TABLE txout
+    (txid TEXT NOT NULL,
+     vout INTEGER NOT NULL,
+     value INTEGER NOT NULL,
+     script BLOB NOT NULL,
+     PRIMARY KEY (txid, vout))
+    STRICT;",
```

Comment:
Aren't both of these the same type: `&indexed_tx_graph::ChangeSet<A, keychain::ChangeSet<K>>`?

What are the differences between these two?

Why not add them to the same table and add a column to differentiate them.
Them everything would be in the `tx` table?

Reply:
These two tables correspond to the `txs` and `txouts` fields of the `tx_graph::ChangeSet<A>` struct and as I understand it are not directly related. The way tx_graph module docs explain it is the `txs` are the "whole" transactions our wallet knows about, and `txouts` are the "partial" transactions where we only know the txid and it's outputs and amounts. 

In tx_graph.rs:
```
pub struct ChangeSet<A = ()> {
    /// Added transactions.
    pub txs: BTreeSet<Transaction>,
    /// Added txouts.
    pub txouts: BTreeMap<OutPoint, TxOut>,
    /// Added anchors.
    pub anchors: BTreeSet<(A, Txid)>,
    /// Added last-seen unix timestamps of transactions.
    pub last_seen: BTreeMap<Txid, u64>,
}
```

Reply:
Okay, I see! Thanks for the explanation!